File: LeetCode_Java/21_validParentheses.java
import java.util.*;

class ValidParentheses{

    public boolean validParentheses(String s){
        Stack<Character> charStack = new Stack<>();
        Map<Character, Character> charMap = new HashMap<>();

        charMap.put(')', '(');
        charMap.put(']', '[');
        charMap.put('}', '{');

        for(char c : s.toCharArray()){
            if(charMap.containsKey(c)){
                if(!charStack.isEmpty() && charStack.peek() == charMap.get(c)){
                    charStack.pop();
                } 
                else{
                    return false;
                }
            } else{
                charStack.push(c);
            }
        }

        return charStack.isEmpty();

    }

    public static void main(String[] args){
        ValidParentheses solution = new ValidParentheses();
        String s = "()[]{}";
        System.out.println(solution.validParentheses(s));
    }  
}

File: LeetCode_Python/21_validParentheses.py
class ValidParentheses:
    def validParentheses(self, s : str) -> bool:
        charStack = []

        charMap = {')' : '(', ']' : '[', '}' : '{' }

        for char in s:
            if char in charMap:
                if charMap and charMap[char] == charStack[-1]:
                    charStack.pop()

                else:
                    return False
            else:
                charStack.append(char)
        
        return True if not charStack else False
    

if __name__ == "__main__":
    solution = ValidParentheses()

    s = "()[]{}"

    print(solution.validParentheses(s))

